src/fluree/db/query/range.cljc

```diff
@@ -59,168 +54,259 @@
+                    (>! error-ch e))))
+    leaf-ch))
+
+(defn flake-range
+  "Returns a channel that will eventually contain a stream of flakes from index
+  `idx` within the database `db` between `start-flake` and `end-flake`,
+  inclusive, one-by-one, and sorted by the order of `idx`. Any exceptions
+  encountered while resolving index nodes will be placed on `error-ch`"
+  [{:keys [conn] :as db}
+   idx
+   {:keys [from-t to-t start-test start-flake
+           end-test end-flake]}
+   error-ch]
+  (let [novelty  (get-in db [:novelty idx])]
+    (-> db
+        (leaf-range idx start-flake end-flake error-ch)
+        (async/pipe (chan 1 (comp (map (fn [leaf]
+                                         (index/at-t leaf to-t novelty)))
+                                  (map :flakes)
+                                  (map (partial index/t-range from-t to-t))
+                                  (mapcat (fn [flakes]
+                                            (flake/subrange flakes
+                                                            start-test start-flake
+                                                            end-test end-flake)))))))))
+
+(defn filter-authorized
+  "Returns a channel that will eventually contain only the schema flakes and the
+  flakes validated by fluree.db.permissions-validate/allow-flake? function for
+  the database `db` from the `flake-stream` channel"
+  [flake-stream {:keys [permissions] :as db} start end error-ch]
+  #?(:cljs
+     flake-stream ; Note this bypasses all permissions in CLJS for now!
+
+     :clj
+     (let [s1 (flake/s start)
+           p1 (flake/p start)
+           s2 (flake/s end)
+           p2 (flake/p end)]
+       (if (perm-validate/no-filter? permissions s1 s2 p1 p2)
+         flake-stream
+         (let [out-ch (chan)]
+           (go
+             (try* (loop []
+                     (when-let [flake (<! flake-stream)]
+                       (when (or (schema-util/is-schema-flake? flake)
+                                 (<? (perm-validate/allow-flake? db flake)))
+                         (>! out-ch flake))
+                       (recur)))
+                   (async/close! out-ch)
+                   (catch* e
+                           (log/error e
+                                      "Error validating flake in ledger"
+                                      (select-keys db [:network :dbid :t]))
+                           (>! error-ch e))))
+           out-ch)))))
+
+(defn flake-filter-xf
+  [{:keys [subject-fn predicate-fn object-fn]}]
+  (let [filter-xfs (cond-> []
+                     subject-fn   (conj (filter (fn [f]
+                                                  (subject-fn (flake/s f)))))
+                     predicate-fn (conj (filter (fn [f]
+                                                  (predicate-fn (flake/p f)))))
+                     object-fn    (conj (filter (fn [f]
+                                                  (object-fn (flake/o f))))))]
```

Comment:
I think we've got room for all these xfs to be on one line each, makes it a bit easier to compare lines.